File: frontend/src/utils/attribution.js
/**
 * Attribution capture — P0 2026-05-22 Phase A.
 *
 * Captures Google Ads / Meta Ads click-IDs and utm_* parameters on
 * the FIRST page load so they survive every downstream redirect
 * (Cashfree hosted-page, OAuth, etc.). Without this, every paid
 * signup and every paid payment loses its attribution at the
 * redirect boundary.
 *
 * Hard requirements (founder mandate 2026-05-22):
 *   • Capture on first page load — not waiting for signup.
 *   • Persist for 90 days (Google Ads click-ID window).
 *   • Send to backend so the cashfree webhook can stamp it on the
 *     order at payment time.
 *   • Source platform classification carried alongside click-IDs.
 *
 * Restraint: this is the entire attribution module. No session
 * replay, no heatmaps, no behavioral scoring. Minimal truth.
 */

const STORAGE_KEY = 'vs_attribution_v1';
const ANON_KEY = 'vs_anonymous_id';
const TTL_MS = 90 * 24 * 60 * 60 * 1000; // 90 days
const POST_TTL_MS = 24 * 60 * 60 * 1000;  // re-POST at most once per day

const CLICK_ID_PARAMS = ['gclid', 'gbraid', 'wbraid', 'fbclid'];
const UTM_PARAMS = ['utm_source', 'utm_medium', 'utm_campaign', 'utm_content', 'utm_term'];

function _safeStorage() {
  try {
    if (typeof window === 'undefined') return null;
    return window.localStorage;
  } catch (_) {
    return null;
  }
}

function _readJson(key) {
  const ls = _safeStorage();
  if (!ls) return null;
  try {
    const raw = ls.getItem(key);
    return raw ? JSON.parse(raw) : null;
  } catch (_) {
    return null;
  }
}

function _writeJson(key, value) {
  const ls = _safeStorage();
  if (!ls) return;
  try {
    ls.setItem(key, JSON.stringify(value));
  } catch (_) { /* quota */ }
}

function _generateAnonymousId() {
  if (typeof window !== 'undefined' && window.crypto && window.crypto.randomUUID) {
    return window.crypto.randomUUID();
  }
  // Fallback: 32-hex random ID.
  let s = '';
  for (let i = 0; i < 32; i++) s += Math.floor(Math.random() * 16).toString(16);
  return s;
}

export function getAnonymousId() {
  const ls = _safeStorage();
  if (!ls) return _generateAnonymousId();
  let id = ls.getItem(ANON_KEY);
  if (!id) {
    id = _generateAnonymousId();
    try { ls.setItem(ANON_KEY, id); } catch (_) { /* quota */ }
  }
  return id;
}

function _classifySourcePlatform(params, referrer) {
  if (params.gclid || params.gbraid || params.wbraid) return 'google_ads';
  if (params.fbclid) return 'meta_ads';
  const utm = (params.utm_source || '').toLowerCase();
  if (utm.includes('google')) return 'google_ads';
  if (['facebook', 'instagram', 'meta'].includes(utm)) return 'meta_ads';
  if (utm) return 'referral';
  if (referrer && referrer.length > 0) return 'referral';
  return 'direct';
}

function _parseUrlAttribution() {
  if (typeof window === 'undefined') return null;
  try {
    const url = new URL(window.location.href);
    const out = {};
    [...CLICK_ID_PARAMS, ...UTM_PARAMS].forEach((p) => {
      const v = url.searchParams.get(p);
      if (v) out[p] = v;
    });
    return out;
  } catch (_) {
    return {};
  }
}

/**
 * Capture attribution from the current page URL and persist it.
 *
 * Idempotency: never overwrites a stronger signal with a weaker one.
 * If localStorage already has a gclid and the current URL has none,
 * we keep the stored gclid (the click-ID window is up to 90 days).
 *
 * Returns the merged stored attribution object (or null if not
 * available, e.g., during SSR).
 */
export function captureAttribution() {
  const ls = _safeStorage();
  if (!ls || typeof window === 'undefined') return null;

  const now = Date.now();
  const urlAttr = _parseUrlAttribution() || {};
  const stored = _readJson(STORAGE_KEY) || {};
  const isExpired = stored.expires_at && stored.expires_at < now;
  const base = isExpired ? {} : (stored.data || {});

  // Merge: URL params win when present; otherwise keep stored.
  const merged = { ...base };
  [...CLICK_ID_PARAMS, ...UTM_PARAMS].forEach((p) => {
    if (urlAttr[p]) merged[p] = urlAttr[p];
  });

  // Stamp landing context only on the FIRST capture (don't overwrite).
  if (!merged.landing_path) merged.landing_path = window.location.pathname;
  if (!merged.referrer && typeof document !== 'undefined') {
    merged.referrer = document.referrer || '';
  }
  merged.source_platform = _classifySourcePlatform(merged, merged.referrer);

  _writeJson(STORAGE_KEY, {
    data: merged,
    expires_at: now + TTL_MS,
    last_captured_at: now,
  });

  return merged;
}

/**
 * Returns the current stored attribution (or null if absent/expired).
 * Pure read — never writes.
 */
export function getStoredAttribution() {
  const stored = _readJson(STORAGE_KEY);
  if (!stored || !stored.data) return null;
  if (stored.expires_at && stored.expires_at < Date.now()) return null;
  return stored.data;
}

/**
 * Send attribution to the backend (best-effort; never throws).
 * Idempotent: re-POSTs at most once per POST_TTL_MS so we don't
 * spam the server on every navigation.
 */
export async function syncAttributionToBackend(apiClient) {
  try {
    const stored = _readJson(STORAGE_KEY);
    if (!stored || !stored.data) return;

    const lastPosted = stored.last_posted_at || 0;
    if (Date.now() - lastPosted < POST_TTL_MS) return;

    const body = {
      anonymous_id: getAnonymousId(),
      ...stored.data,
    };
    await apiClient.post('/api/attribution/capture', body);

    _writeJson(STORAGE_KEY, {
      ...stored,
      last_posted_at: Date.now(),
    });
  } catch (_) {
    // Network errors must never break the app. The next page load
    // will re-attempt.
  }
}

File: backend/routes/attribution.py
"""
Attribution capture endpoint — P0 2026-05-22 Phase A.

Persists Google Ads / Meta Ads / utm_* click-IDs from the FIRST page
load so they survive every downstream redirect (Cashfree hosted-page,
OAuth, etc.). Without this, every paid signup and every paid payment
loses its attribution at the redirect boundary.

Hard requirements (founder mandate 2026-05-22):
  • Capture on first page load — not waiting for signup.
  • Idempotent for the same anonymous_id (we never overwrite a real
    gclid with a missing one on subsequent navigations).
  • Carries source_platform alongside the click-IDs so downstream
    funnel analysis can slice by paid channel without a join.

Doctrine refs:
  • rule 1 (validate every boundary) — strict Pydantic envelope
  • rule 5 (no leaked internals) — only writes; no read endpoint exposed
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel, Field

from shared import db, logger, get_optional_user

router = APIRouter(prefix="/attribution", tags=["attribution"])

# ─── Allow-listed source platforms ──────────────────────────────────
# Strict literal-style validation: we don't want free-form strings
# poisoning analytics downstream. New platforms add here deliberately.
ALLOWED_SOURCE_PLATFORMS = frozenset({
    "google_ads", "meta_ads", "twitter_ads", "linkedin_ads",
    "tiktok_ads", "youtube_ads",
    "organic", "direct", "referral", "email", "unknown",
})


class AttributionCapturePayload(BaseModel):
    anonymous_id: str = Field(..., min_length=8, max_length=64)
    gclid: Optional[str] = Field(default=None, max_length=512)
    gbraid: Optional[str] = Field(default=None, max_length=512)
    wbraid: Optional[str] = Field(default=None, max_length=512)
    fbclid: Optional[str] = Field(default=None, max_length=512)
    utm_source: Optional[str] = Field(default=None, max_length=200)
    utm_medium: Optional[str] = Field(default=None, max_length=200)
    utm_campaign: Optional[str] = Field(default=None, max_length=200)
    utm_content: Optional[str] = Field(default=None, max_length=200)
    utm_term: Optional[str] = Field(default=None, max_length=200)
    landing_path: Optional[str] = Field(default=None, max_length=2048)
    referrer: Optional[str] = Field(default=None, max_length=2048)
    source_platform: Optional[str] = Field(default=None, max_length=32)


def _classify_source_platform(p: AttributionCapturePayload) -> str:
    """Derive a canonical source_platform from the strongest signal
    available. Order matters: paid click-IDs win over utm_* over
    referrer over the caller-supplied value over 'direct'."""
    if p.gclid or p.gbraid or p.wbraid:
        return "google_ads"
    if p.fbclid:
        return "meta_ads"
    sp = (p.source_platform or "").lower()
    if sp in ALLOWED_SOURCE_PLATFORMS:
        return sp
    utm = (p.utm_source or "").lower()
    if "google" in utm:
        return "google_ads"
    if utm in ("facebook", "instagram", "meta"):
        return "meta_ads"
    if utm:
        return "referral"
    if p.referrer:
        return "referral"
    return "direct"


@router.post("/capture")
async def capture_attribution(
    payload: AttributionCapturePayload,
    request: Request,
    user: Optional[dict] = Depends(get_optional_user),
):
    """Persist attribution against an anonymous session id.

    Idempotency: we never overwrite an existing real gclid/utm field
    with an empty one. Each call merges only the fields that are
    actually present and non-empty in the payload.
    """
    now_iso = datetime.now(timezone.utc).isoformat()
    source_platform = _classify_source_platform(payload)

    set_fields: dict = {
        "anonymous_id": payload.anonymous_id,
        "last_seen_at": now_iso,
        "source_platform": source_platform,
    }
    if user and user.get("id"):
        set_fields["user_id"] = user["id"]

    # Only merge fields that are present and non-empty. This is the
    # idempotency guarantee: subsequent visits without a gclid never
    # clobber the original capture.
    for fld in ("gclid", "gbraid", "wbraid", "fbclid",
                "utm_source", "utm_medium", "utm_campaign",
                "utm_content", "utm_term",
                "landing_path", "referrer"):
        v = getattr(payload, fld, None)
        if v:
            set_fields[fld] = v

    await db.attribution_sessions.update_one(
        {"anonymous_id": payload.anonymous_id},
        {
            "$set": set_fields,
            "$setOnInsert": {
                "first_seen_at": now_iso,
                "first_landing_path": payload.landing_path or "",
                "first_referrer": payload.referrer or "",
            },
        },
        upsert=True,
    )

    request_id = getattr(request.state, "request_id", None)
    logger.info(
        "[ATTRIBUTION] captured anonymous_id=%s source=%s gclid=%s utm_source=%s request_id=%s",
        payload.anonymous_id, source_platform,
        bool(payload.gclid), payload.utm_source or "-", request_id,
    )

    return {
        "ok": True,
        "source_platform": source_platform,
        "request_id": request_id,
    }


async def get_attribution_for_user(user_id: str) -> dict:
    """Look up the most recent attribution session for a given user.

    Used by the cashfree webhook to stamp the order with the paid-click
    context so the conversion fire has the correct source_platform.
    Returns {} when not found (safe-default).
    """
    if not user_id:
        return {}
    doc = await db.attribution_sessions.find_one(
        {"user_id": user_id},
        sort=[("last_seen_at", -1)],
        projection={"_id": 0},
    )
    return doc or {}
